functime/conformal.py

```diff
@@ -1,76 +1,109 @@
 from __future__ import annotations
 
-from typing import List, Optional
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Optional, Sequence, Union
 
 import polars as pl
 
 
-def enbpi(
-    y_pred: pl.LazyFrame,
-    y_resid: pl.LazyFrame,
-    alphas: List[float],
+def conformalize(
+    *,
+    y_pred: Union[pl.DataFrame, pl.LazyFrame],
+    y_preds: Union[pl.DataFrame, pl.LazyFrame],
+    y_resids: Union[pl.DataFrame, pl.LazyFrame],
+    alphas: Optional[Sequence[float]] = None,
+) -> pl.DataFrame:
+    """Compute prediction intervals using ensemble batch prediction intervals (ENBPI).
+
+    Parameters
+    ----------
+    y_pred : Union[pl.DataFrame, pl.LazyFrame]
+        The predicted values.
+    y_preds : Union[pl.DataFrame, pl.LazyFrame]
+        The predictions resulting from backtesting.
+    y_resids : Union[pl.DataFrame, pl.LazyFrame]
+        The backtesting residuals.
+    alphas : Optional[Sequence[float]]
+        The quantile levels to use for the prediction intervals. Defaults to (0.1, 0.9).
+        Quantiles must be two values between 0 and 1 (exclusive).
+
+    Returns
+    -------
+    pl.DataFrame
+        The prediction intervals.
+    """
+    alphas = _validate_alphas(alphas)
+
+    entity_col, time_col, target_col = y_pred.columns[:3]
+    schema = y_pred.schema
+
+    _y_resids: pl.LazyFrame = y_resids.lazy().select(y_resids.columns[:3])
+    _y_preds: pl.LazyFrame = pl.concat(
+        [
+            y_pred.lazy(),
+            y_preds.lazy().select(
+                entity_col,
+                pl.col(time_col).cast(schema[time_col]),
+                pl.col(target_col).cast(schema[target_col]),
+            ),
+        ]
+    )
+
+    y_pred_quantiles = _compute_enbpi(
+        y_preds=_y_preds,
+        y_resids=_y_resids,
+        alphas=alphas,
+    )
+
+    # Make alpha base 100
+    y_pred_quantiles = y_pred_quantiles.with_columns(
+        (pl.col("quantile") * 100).cast(pl.Int16)
```

Comment:
If you have more than 2 decimal places, then the casting would just truncate the decimal values 🤔

Reply:
That's correct; I am still to change that bit of the code fortunately. This column won't exist once I'm done ✔️